feat: dockge set/update agent friendly name

lohrbini opened the pull request:
⚠️⚠️⚠️ Since we do not accept all types of pull requests and do not want to waste your time. Please be sure that you have read pull request rules:
https://github.com/louislam/dockge/blob/master/CONTRIBUTING.md

Tick the checkbox if you understand [x]: 
- [x] I have read and understand the pull request rules.

# Disclaimer
I am not a Software programmer and only have basic understandings on any Programming Languages used in the project, so please be patient and maybe assist me with further development of this feature

# Description
Since `dockge agents` is still a beta feature and in the future maybe more and more remote instances will be added I wanted to enable the option for setting a friendly name for a remote instance. If the name is not set via the prompt. The default `agent.url` will be displayed
Fixes #345

## Type of change
- New feature (non-breaking change which adds functionality)


- [x] My code follows the style guidelines of this project
- [x] I ran ESLint and other linters for modified files
- [x] I have performed a self-review of my own code and tested it
- [x] I have commented my code, particularly in hard-to-understand areas
  (including JSDoc for methods)
- [x] My changes generate no new warnings
- [ ] My code needed automated testing. I have added them (this is optional task)

## Screenshots 

![dgagents-friendlyname](https://github.com/louislam/dockge/assets/10088089/6be5ff31-c4cc-4db5-b882-6abc21b79e3a)

lohrbini committed:
feat: dockge agent friendly name

Add the option to set a friendly name to a dockge remote
agent. If the field is not used the `agent.url` will be
displayed

cyril59310 commented:
After testing your PR, I have two small remarks:

1) Could you add the translation key "Friendly Name" to the en.json file?
2) It would be interesting for the friendly name to be applied here as well
![si7sb7](https://github.com/louislam/dockge/assets/70776486/5f7e3d2f-a3ec-4f85-a224-d82dae1ba5f4)

cyril59310 commented:
In the file 

> frontend/src/lang/en.json ,

please add `"Friendly Name": "Friendly Name"` so that translators can translate it into other languages using our Weblate tool.

lohrbini committed:
docs(en): add missing new item

adding `Friendly Name` to en.json

lohrbini committed:
DRAFT: feat(friendlyName): make friendlyname configurable

allow updates and removal of `friendly name`

lohrbini commented:
@cyril59310 bulletpoint 2 of your request is now already done. Only need to update codebase and remove all debug options

![image](https://github.com/louislam/dockge/assets/10088089/269d56db-4988-43aa-80f2-64fcd97e3da0)

lohrbini committed:
feat(containerTerminal): add friendly name

adding the `friendlyname` option to the running and paused container

larswmh commented on frontend/src/mixins/socket.ts:
```diff
@@ -131,10 +131,15 @@ export default defineComponent({
-                return this.$t("currentEndpoint");
+            for (const [k, v] of Object.entries(this.$data.agentList)) {
+                if (endpoint) {
+                    if (endpoint === v['endpoint'] && v['friendlyname'] !== "") {
+                        return v['friendlyname'];
+                    }
+                    if (endpoint === v['endpoint'] &&  v['friendlyname'] === "" ) {
+                        return endpoint;
+                    }
```
Thank you so much for this pull request. Just a small suggestion to avoid [CI test failure](https://github.com/louislam/dockge/actions/runs/7748302986/job/21130673284?pr=414#step:10:94)
- Replaces single quotes with double quotes
- Remove duplicate space in Line 139
```suggestion
                    if (endpoint === v["endpoint"] && v["friendlyname"] !== "") {
                        return v["friendlyname"];
                    }
                    if (endpoint === v["endpoint"] && v["friendlyname"] === "") {
                        return endpoint;
                    }
```

lohrbini replied:
Oh sorry! Did not see your suggestions I already rebased. But thanks in advance!

cyril59310 commented:
Thank you for taking my feedback. 
I noticed afterwards that here too, the friendly name could be added.
![image](https://github.com/louislam/dockge/assets/70776486/d6af505c-e407-43bf-979e-b4b12c12c43c)

lohrbini committed:
feat(compose): add friendly names

when create a new stack via UI the friendly name should also be displayed

lohrbini commented:
Thanks again @cyril59310 . This is now also implemented

lohrbini committed:
chore(deps): update from upstream

Merge branch 'feat/dockge-agents-friendly-name' into feat/dockge-agents-friendly-name-editable

lohrbini committed:
fix(agent-manager): add update query

add a query to update the friendlyname correctly

lohrbini committed:
fix(DashboardHome): remove edit-pen for controller

remove edit-pen for controller which cannot have a friendlyname for now

lohrbini commented:
Finally I was able to get the `update` feature working. Friendly names are now updated properly by a small dialog

lohrbini committed:
style(agent-manager): liniting

fix linting issues

lohrbini committed:
fix(manage-agent-socket-handler): type

fix tsc check by updating var type

lohrbini committed:
fix(DashboardHome): duplicate code

remove if condition and update to proper values to prevent
duplicates

cyril59310 commented:
J'ai fait quelques suggestions afin d'ajouter des clés de traduction pour nos traducteurs.

louislam commented on backend/migrations/2023-12-20-2117-agent-table.ts:
```diff
@@ -7,6 +7,7 @@ export async function up(knex: Knex): Promise<void> {
         table.string("url", 255).notNullable().unique();
         table.string("username", 255).notNullable();
         table.string("password", 255).notNullable();
+        table.string("friendlyname", 255);
```
This file cannot be modified. Have to create a new file. 

Migration guide:
https://github.com/louislam/uptime-kuma/tree/master/db/knex_migrations


Will come back later, as I would like to finish the next milestone of Uptime Kuma first.

louislam commented on backend/agent-manager.ts:
```diff
@@ -76,12 +76,15 @@ export class AgentManager {
+     * @param updatedFriendlyName
      */
-    async add(url : string, username : string, password : string) : Promise<Agent> {
+    async add(url : string, username : string, password : string, friendlyname : string) : Promise<Agent> {
```
Name Conventions

- Javascript/Typescript: camelCaseType
- SQLite: snake_case (Underscore)
- CSS/SCSS: kebab-case (Dash)


https://github.com/louislam/dockge/blob/master/CONTRIBUTING.md#name-conventions

lohrbini replied:
all namings has been updated

nathan815 commented on backend/agent-manager.ts:
```diff
@@ -104,6 +107,16 @@ export class AgentManager {
+     */
+
+    async update(friendlyname : string, updatedFriendlyName : string) {
+        await R.exec("UPDATE agent SET friendlyname = " + updatedFriendlyName + " WHERE friendlyname = " + friendlyname + "");
```
SQL injection vuln here - if name has quotes this will break.

You can use parameter bindings to build queries safely: 

```ts
await R.exec("UPDATE agent SET friendlyname = ? WHERE friendlyname = ?", [
    updatedFriendlyName,
    friendlyname
]);
```

https://github.com/louislam/redbean-node/blob/master/docs/Query.md

nathan815 replied:
Instead of a raw query, you could utilize the redbean ORM methods. See: https://github.com/louislam/redbean-node/blob/master/docs/Create-Update-Bean.md

Something like this if by ID (primary key):

```ts
const agent = await R.load('agent', id);
agent.friendlyName = friendlyName;
await R.store(agent);
```

Or by URL:

```ts
const agent = await R.findOne('agent', ' url = ? ', [url]);
agent.friendlyName = friendlyName;
await R.store(agent);
```

lohrbini replied:
Thanks for the input. Added this with `update by url` and started a small "discussion"
https://github.com/louislam/dockge/pull/414#discussion_r1705773795

nathan815 commented on backend/agent-manager.ts:
```diff
@@ -104,6 +107,16 @@ export class AgentManager {
+     * @param updatedFriendlyName
+     */
+
+    async update(friendlyname : string, updatedFriendlyName : string) {
```
I would consider accepting the agent's `id` or the `url` here for identifying the agent to update instead of friendly name.

ID is the primary key, so thats probably best to use for updating the agent. But URL is always set and has a unique constraint, so it will work too.

kayakyakr replied:
I agree with nathan, I would refactor to pass id around for the agent

lohrbini replied:
sure it should be the `id` ? it is currently not passed in the agentList and needs to be added. If a host will be added and deleted it will still increase and not release a taken number. The `url` would be the "better" identifier ? 
( just a small comment. I am not against refactoring this to `id` but while updating this MR I was thinking about)

nathan815 commented on backend/models/agent.ts:
```diff
@@ -23,6 +23,8 @@ export class Agent extends BeanModel {
             url: this.url,
             username: this.username,
             endpoint: this.endpoint,
+            friendlyname: this.friendlyname,
+            updatedFriendlyName: this.updatedFriendlyName
```
Why is `updatedFriendlyName` also here on the model? Shouldn't just `friendlyName` be sufficient?

nathan815 reviewed (commented):
Nice feature!

Have some code feedback

lohrbini commented:
Long time no see. Will have a look the next days on all your feedback. Many thanks 🙏🏻

kayakyakr commented on backend/models/agent.ts:
```diff
@@ -23,6 +23,8 @@ export class Agent extends BeanModel {
             url: this.url,
             username: this.username,
             endpoint: this.endpoint,
+            friendlyname: this.friendlyname,
```
recommendation: you should consider just naming this field `name`. There are no conflicts using the simpler version of the field name.

kayakyakr commented on backend/socket-handlers/manage-agent-socket-handler.ts:
```diff
@@ -66,5 +66,31 @@ export class ManageAgentSocketHandler extends SocketHandler {
+                log.debug("manage-agent-socket-handler", "updateAgent");
+                checkLogin(socket);
+
+                if (typeof(updatedFriendlyName) !== "string") {
+                    throw new Error("FriendlyName must be a string");
+                }
```
the more you know: since this is typescript, you shouldn't actually need to typecheck the input value for friendly name. We already know it's a string from before.

kayakyakr commented on frontend/src/pages/DashboardHome.vue:
```diff
@@ -49,13 +49,25 @@
+                            </template>
+
+                            <!-- Edit FriendlyName  -->
+                            <font-awesome-icon v-if="agent.friendlyname !== '' && agent.friendlyname !== ''" icon="pen-to-square" @click="showEditAgentFriendlynameDialog[agent.friendlyname] = !showEditAgentFriendlynameDialog[agent.friendlyname]" />
```
suggestion: I'm not sure why you're checking for blank string twice here, but if you switch to passing around `agent.id` you will be able to edit a name no matter if the server was created with friendly name or not, so the v-if will be unnecessary.

kayakyakr commented on frontend/src/pages/DashboardHome.vue:
```diff
@@ -49,13 +49,25 @@
+                            <!-- Edit Dialog -->
+                            <BModal v-model="showEditAgentFriendlynameDialog[agent.friendlyname]" :no-close-on-backdrop="true" :close-on-esc="true" :okTitle="$t('Update Friendlyname')" okVariant="info" @ok="updateFriendlyname(agent.friendlyname, agent.updatedFriendlyName)">
+                                <label for="Update Friendlyname" class="form-label">Current value: {{ $t(agent.friendlyname) }}</label>
+                                <input id="updatedFriendlyName" v-model="agent.updatedFriendlyName" type="text" class="form-control" optional>
```
suggestion: similar to nathan's commends up above, if you are finding agents by their ID, you can just set `agent.name` here and use that to update the value. You also then don't have to show `Current value:` in the label, because the input will have the current value. You may want to show `agent.url` as part of the label, though, in case input name is blank.

kayakyakr commented on frontend/src/pages/DashboardHome.vue:
```diff
@@ -49,13 +49,25 @@
                             <!-- Agent Display Name -->
-                            <span v-if="endpoint === ''">{{ $t("currentEndpoint") }}</span>
-                            <a v-else :href="agent.url" target="_blank">{{ endpoint }}</a>
+                            <template v-if="$root.agentStatusList[endpoint]">
+                                <span v-if="endpoint === '' && agent.friendlyname === ''" class="badge bg-secondary me-2">Controller</span>
+                                <span v-else-if="agent.friendlyname === ''" :href="agent.url">{{ endpoint }}</span>
+                                <span v-else :href="agent.url">{{ agent.friendlyname }}</span>
+                            </template>
```
Couple of suggestions here: 

1. We can use the fact that an empty string is falsy to shorten this chain.
2. You probably don't want to be modifying things so drastically. If it's working, then a small change will do.

```suggestion
                            <span v-if="endpoint === ''" class="badge bg-secondary me-2">{{ $t("currentEndpoint") }}</span>
                            <a v-else :href="agent.url" target="_blank">{{ agent.name || endpoint }}</a>
```

kayakyakr reviewed (commented):
I appreciate this addition. Short of hacking the docker swarm mechanics to do node auto-discovery, I think is a fairly nice quality of life improvement.

Left some comments on some changes I'd make to make this prod-ready.

mon5termatt commented:
would love for this to be added!

lohrbini committed:
fix: pull-request suggestions

updating the merge request by adding the suggestions from reviewer

lohrbini committed:
fix(suggestion): added missing change

update agent.ts file by renaming the variables and added suggestion

lohrbini commented:
Ready & happy about a second review :)

lohrbini committed:
chore(lang): updated en.json

updated variables in language file

larswmh commented on frontend/src/pages/DashboardHome.vue:
```diff
@@ -49,13 +49,25 @@
-                            <span v-if="endpoint === ''">{{ $t("currentEndpoint") }}</span>
-                            <a v-else :href="agent.url" target="_blank">{{ endpoint }}</a>
+                            <template v-if="$root.agentStatusList[endpoint]">
+                                <span v-if="endpoint === '' && agent.name === ''" class="badge bg-secondary me-2">Controller</span>
+                                <span v-else-if="agent.name === ''" :href="agent.url">{{ endpoint }}</span>
+                                <span v-else :href="agent.url">{{ agent.name }}</span>
```
```suggestion
                                <span v-if="endpoint === '' && agent.name === ''" class="badge bg-secondary me-2">Controller</span>
                                <span v-else-if="agent.name === ''" :href="agent.url">{{ endpoint }}</span>
                                <span v-else :href="agent.url">{{ agent.name }}</span>
```

lohrbini replied:
Thanks for the feedback. Updated in latest commit

larswmh commented on backend/agent-manager.ts:
```diff
@@ -76,12 +76,15 @@ export class AgentManager {
      * @param username
      * @param password
+     * @param name
+     * @param updatedName
```
```suggestion
     * @param updatedName
```
Parameter updatedName is not used in this function, can be removed

lohrbini replied:
Thanks for the feedback. Updated in latest commit

XenoUniv3rse commented:
on which tag do i need to be to have this feature?

lohrbini commented:
> on which tag do i need to be to have this feature? 
> 
> 

It is planned to be included in the next release v1.5.0 . To use this feature you need to be on my branch and build the image on your own

lohrbini committed:
fix: suggestions

implement suggestions from review

lohrbini committed:
Merge branch 'master' into feat/dockge-agents-friendly-name

lohrbini committed:
fix(style): missing comma

add missing comma in en.json

Dracrius reviewed (changes requested):
With these two changes we can wipe out the obvious bugs hanging around from the edits.

![image](https://github.com/user-attachments/assets/b5e4938a-ce51-494d-a182-39fc71c3c5ba)
![image](https://github.com/user-attachments/assets/8af27cf4-053a-4b84-8587-f94a8414e1e1)

I'll also say I much prefer the original name of `Current` for the current agent. I don't really understand the change to `Controller`.

cmcooper1980 committed:
Update agent display name in Compose.vue

commits by @Dracrius

cmcooper1980 committed:
Update agent display name logic in DashboardHome.vue

commits by @Dracrius
